src/countBy.test-d.ts

```diff
@@ -0,0 +1,35 @@
+import { countBy } from "./countBy";
+import { pipe } from "./pipe";
+
+describe("dataFirst", () => {
+  test("countBy", () => {
+    const data = ["a", "b", "c", "B", "A", "a"];
+    const result = countBy(data, (x) => x.toLowerCase());
+
+    expectTypeOf(result).toEqualTypeOf<Record<string, number>>();
+  });
+});
+
+describe("dataLast", () => {
+  test("countBy", () => {
+    const data = ["a", "b", "c", "B", "A", "a"];
+    const result = pipe(
+      data,
+      countBy((x) => x.toLowerCase()),
+    );
+
+    expectTypeOf(result).toEqualTypeOf<Record<string, number>>();
+  });
+});
+
+test("callback function type does not match data array type", () => {
+  const data = [1, 2, 3];
+  // @ts-expect-error [ts2345]: Type 'number' is not assignable to type 'string'.
+  countBy(data, (x: string) => x);
+});
+
+test("array with nullish values", () => {
+  const data = [1, null, undefined];
+  // @ts-expect-error [ts2322] - Type 'number | null | undefined' is not assignable to type 'PropertyKey'.
+  countBy(data, (x) => x);
+});
```

Comment:
This test isn't valuable. It incorrectly states that it tests arrays with nullish values, but those aren't actually the thing being tested (e.g. `countBy([null], (x) => 1)` would work just like any other value would), and what it does test is that the callback function must return a PropertyKey, and that's already enforced by the typing of the function.

```suggestion
```